sql/catalyst/src/main/scala/org/apache/spark/sql/types/DecimalType.scala

```diff
@@ -94,12 +94,15 @@ case class DecimalType(precision: Int, scale: Int) extends FractionalType {
    */
   private[sql] def isTighterThan(other: DataType): Boolean = isTighterThanInternal(other)
 
-  @tailrec
   private def isTighterThanInternal(other: DataType): Boolean = other match {
     case dt: DecimalType =>
       (precision - scale) <= (dt.precision - dt.scale) && scale <= dt.scale
     case dt: IntegralType =>
-      isTighterThanInternal(DecimalType.forType(dt))
+      val integerAsDecimal = DecimalType.forType(dt)
+      assert(integerAsDecimal.scale == 0)
+      // If the precision equals `integerAsDecimal.precision`, there can be integer overflow
+      // during casting.
+      precision < integerAsDecimal.precision && scale == 0
```

Comment:
good catch ! 

it seems we can combine the methods `isTighterThan` and `isTighterThanInternal` since there is no recursion.

Reply:
Sure, done.